src/plot_api/plot_api.js

```diff
@@ -205,7 +205,7 @@ Plotly.plot = function(gd, data, layout, config) {
             if(trace.visible !== true || !trace._module.colorbar) {
                 Plots.autoMargin(gd, 'cb' + trace.uid);
             }
-            else trace._module.colorbar(gd, cd);
+            else if(trace.marker.colorbar) trace._module.colorbar(gd, cd);
```

Comment:
This new `if()` makes traces that have root-level `colorbar` settings (e.g. `heatmap`, `contour`, `surface`) bypass the colorbar draw step. 

Something like

```js
else if(Plots.traceIs(trace, 'markerColorscale') && Colorbar.hasColorbar(trace)) {
  trace._module.colorbar(gd, cd);
}
```

might make things work.